### PR TITLE
Fold same-typed variables if there are more than 3

### LINKED ISSUE #3701: Fold same-typed variables if there are more than 3
Take the example when we have many variables that they occupy so much screen space:
```
[0x0001ee34]> pdf
Do you want to print 258 lines? (y/N) y
            ; CALL XREF from fcn.0001cb0c @ 0x1cdc8
╭ fcn.0001ee34(int64_t arg1, int64_t arg2);
│           ; arg int64_t arg1 @ x0
│           ; arg int64_t arg2 @ x1
│           ; var int64_t var_394h @ stack - 0x394
│           ; var int64_t var_390h @ stack - 0x390
│           ; var int64_t var_300h @ stack - 0x300
│           ; var int64_t var_2f0h @ stack - 0x2f0
│           ; var int64_t var_2e0h @ stack - 0x2e0
│           ; var int64_t var_2d0h @ stack - 0x2d0
│           ; var int64_t var_2c0h @ stack - 0x2c0
│           ; var int64_t var_2b0h @ stack - 0x2b0
│           ; var int64_t var_294h @ stack - 0x294
│           ; var int64_t var_290h @ stack - 0x290
│           ; var int64_t var_200h @ stack - 0x200
│           ; var int64_t var_1f0h @ stack - 0x1f0
│           ; var int64_t var_1e0h @ stack - 0x1e0
│           ; var int64_t var_1d0h @ stack - 0x1d0
│           0x0001ee34      7f2303d5       pacibsp
│           0x0001ee38      ff8303d1       sub   sp, sp, 0xe0
│           0x0001ee3c      f85f0aa9       stp   x24, x23, [var_1d0h + 0x190]
│           0x0001ee40      f6570ba9       stp   x22, x21, [var_1d0h + 0x1a0]
│           0x0001ee44      f44f0ca9       stp   x20, x19, [var_1d0h + 0x1b0]
│           0x0001ee48      fd7b0da9       stp   x29, x30, [var_1d0h + 0x1c0]
│           0x0001ee4c      fd430391       add   x29, sp, 0xd0
│           0x0001ee50      1f2003d5       nop
```
We could try group same-typed variables if there are 3 or more:
```
[0x0001ee34]> pdf
Do you want to print 258 lines? (y/N) y
            ; CALL XREF from fcn.0001cb0c @ 0x1cdc8
╭ fcn.0001ee34(int64_t arg1, int64_t arg2);
│           ; arg int64_t arg1 @ x0
│           ; arg int64_t arg2 @ x1
│           ; var int64_t [var_394h @ stack - 0x394; var_390h @ stack - 0x390;  var_300h @ stack - 0x300; var_2f0h @ stack - 0x2f0 ]
│           ; var int64_t [ var_2e0h @ stack - 0x2e0; var_2d0h @ stack - 0x2d0; var_2c0h @ stack - 0x2c0; var_2b0h @ stack - 0x2b0 ]
│           ; var int64_t [ var_294h @ stack - 0x294; var_290h @ stack - 0x290; var_200h @ stack - 0x200; var_1f0h @ stack - 0x1f0 ]
│           ; var int64_t var_1e0h @ stack - 0x1e0
│           ; var int64_t var_1d0h @ stack - 0x1d0
│           0x0001ee34      7f2303d5       pacibsp
│           0x0001ee38      ff8303d1       sub   sp, sp, 0xe0
│           0x0001ee3c      f85f0aa9       stp   x24, x23, [var_1d0h + 0x190]
│           0x0001ee40      f6570ba9       stp   x22, x21, [var_1d0h + 0x1a0]
│           0x0001ee44      f44f0ca9       stp   x20, x19, [var_1d0h + 0x1b0]
│           0x0001ee48      fd7b0da9       stp   x29, x30, [var_1d0h + 0x1c0]
│           0x0001ee4c      fd430391       add   x29, sp, 0xd0
│           0x0001ee50      1f2003d5       nop
```
We could have 2 different levels of folding, e.g. with `e scr.var.fold` option, the upper one is 1, and if set to 2 then we just show `...` for all that didn't fit into the screen:
```
[0x0001ee34]> pdf
Do you want to print 258 lines? (y/N) y
            ; CALL XREF from fcn.0001cb0c @ 0x1cdc8
╭ fcn.0001ee34(int64_t arg1, int64_t arg2);
│           ; arg int64_t arg1 @ x0
│           ; arg int64_t arg2 @ x1
│           ; var int64_t [var_394h @ stack - 0x394; var_390h @ stack - 0x390;  ... ]
│           0x0001ee34      7f2303d5       pacibsp
│           0x0001ee38      ff8303d1       sub   sp, sp, 0xe0
│           0x0001ee3c      f85f0aa9       stp   x24, x23, [var_1d0h + 0x190]
│           0x0001ee40      f6570ba9       stp   x22, x21, [var_1d0h + 0x1a0]
│           0x0001ee44      f44f0ca9       stp   x20, x19, [var_1d0h + 0x1b0]
│           0x0001ee48      fd7b0da9       stp   x29, x30, [var_1d0h + 0x1c0]
│           0x0001ee4c      fd430391       add   x29, sp, 0xd0
│           0x0001ee50      1f2003d5       nop

### PR DESCRIPTION
**Pull request by PeiweiHu**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

There are three choice for `asm.var.fold`:
+ none -> no fold
+ group -> fold three variables in each group
+ hide -> group the first two variables with ellipsis in tail

as shown in the following picture.

image

**Test plan**

...

**Closing issues**

closes https://github.com/rizinorg/rizin/issues/3701

### CODE REVIEW

**Comment by ret2libc** [comment]
Minor thing, but instead of the various 1,2,3,etc. we were used to use in many cases, I think it makes more sense to use options.

See for example:
```c
	SETDESC(n, "Select assembly syntax");
	SETOPTIONS(n, "att", "intel", "masm", "jz", "regnum", NULL);
```
we could use `asm.var.fold=none`, `asm.var.fold=group`, `asm.var.fold=hide`. Or other values that you think fit better :)

File: librz/core/disasm.c [hunk 1]
```diff
@@ -1839,25 +1841,82 @@ static void ds_show_functions(RzDisasmState *ds) {
...
 			printVarSummary(ds, all_vars);
-		} else {
-			char spaces[32];
+		} else { // show_varum = 0 and -1
 			RzAnalysisVar *var;
 			RzListIter *iter;
 			RzList *all_vars = vars_cache.regvars;
 			rz_list_join(all_vars, vars_cache.stackvars);
 			rz_list_foreach (all_vars, iter, var) {
+				if (ds->fold_var && !rz_analysis_var_is_arg(var)) {
```
Comment (XVilka) [inline librz/core/disasm.c]: I suggest extracting that in the separate function

File: librz/core/disasm.c [hunk 2]
```diff
@@ -1839,25 +1841,82 @@ static void ds_show_functions(RzDisasmState *ds) {
...
+					char *vartype = rz_type_as_string(core->analysis->typedb, var->type);
+					RzListIter *temp_it = iter->n;
+					ut32 same_type_cnt = 1;
+					while (temp_it) {
+						RzAnalysisVar *temp_var = temp_it->data;
+						char *temp_vartype = rz_type_as_string(core->analysis->typedb, temp_var->type);
```
Comment (XVilka) [inline librz/core/disasm.c]: Beware of the NULL here, both `temp_var` and `temp_vartype` could be that.

**Review by XVilka — CHANGES_REQUESTED** [review]
Agree with @ret2libc. The output looks good also.

**Comment by PeiweiHu** [comment]
Thanks for the review. The pr is changed as suggested :)

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]

**Pull request merged**